File: Cpp/Logger/LogConfig.cpp
//---------------------------------------------------//
//                    MIT License                    //
// Copyright @ 2018-2019 Tony Su All Rights Reserved //
//        https://github.com/peitaosu/Logger         //
//---------------------------------------------------//

#include "LogConfig.h"

LogConfig::LogConfig(){
    LogConfig("LogConfig.xml");
}

LogConfig::LogConfig(std::string config) {
    this->config = config;
    tinyxml2::XMLDocument xml_doc;
    if (xml_doc.LoadFile(this->config.c_str()) != 0)
    {
        std::cout << "Failed to load data from XML: " << config.c_str() << std::endl;
        return;
    }
    tinyxml2::XMLNode* root = xml_doc.FirstChildElement("Logger");
    for (tinyxml2::XMLElement* element = root->FirstChildElement(); element != NULL; element = element->NextSiblingElement())
    {
        if (strcmp(element->Name(), "Appender") == 0) {
            LogAppender appender;
            appender.Enabled = false;
            appender.Name = element->Attribute("Name");
            appender.Type = element->Attribute("Type");
            for (tinyxml2::XMLElement* appenderElement = element->FirstChildElement(); appenderElement != NULL; appenderElement = appenderElement->NextSiblingElement()) {
                if (strcmp(appenderElement->Name(), "File") == 0) {
                    appender.File = { appenderElement->Attribute("Path"), appenderElement->Attribute("AppendTo") };
                }else if (strcmp(appenderElement->Name(), "Pattern") == 0) {
                    appender.Pattern = { appenderElement->Attribute("Value") };
                }else if (strcmp(appenderElement->Name(), "Color") == 0) {
                    AppenderColor color;
                    color.Level = appenderElement->Attribute("Level");
                    if (appenderElement->Attribute("ForeColor")) color.ForeColor = appenderElement->Attribute("ForeColor");
                    if (appenderElement->Attribute("BackColor")) color.ForeColor = appenderElement->Attribute("BackColor");
                    appender.Colors.push_back(color);
                }
            }
            this->_logAppenders.push_back(appender);
        }
        else if(strcmp(element->Name(), "Root") == 0){
            for (tinyxml2::XMLElement* rootElement = element->FirstChildElement(); rootElement != NULL; rootElement = rootElement->NextSiblingElement()) {
                if (strcmp(rootElement->Name(), "MinLevel") == 0) {
                    this->_logRoot.MinLevel = { rootElement->Attribute("Value") };
                }else if (strcmp(rootElement->Name(), "AppenderRef") == 0) {
                    this->_logRoot.RootAppenderRefs.push_back({ rootElement->Attribute("Ref") });
                }
            }
        }
    }
    for (auto& appender : this->_logAppenders) {
        for (auto& ref : this->_logRoot.RootAppenderRefs) {
            if (appender.Name.compare(ref.Ref) == 0) {
                appender.Enabled = true;
            }
        }
    }
}

LogConfig::~LogConfig(){

}

void LogConfig::UpdateLogAppenders(std::list<LogAppender> appenders) {
    this->_logAppenders = appenders;
}

std::list<LogAppender> LogConfig::GetLogAppenders()
{
    return this->_logAppenders;
}

LogRoot LogConfig::GetLogRoot()
{
    return this->_logRoot;
}

File: CSharp/Logger/LogConfig.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Serialization;

namespace Logging
{
    [XmlRoot(ElementName = "Logger", Namespace = "http://Logger/LogConfig.xsd")]
    public class LogConfig
    {
        [XmlElement("Appender")]
        public List<LogAppender> Appenders;
        [XmlElement("Root")]
        public LogRoot Root;

        public LogConfig()
        {
            Appenders = new List<LogAppender>();
        }

        public struct LogAppender
        {
            [XmlAttribute]
            public string Name;
            [XmlAttribute]
            public string Type;
            [XmlElement("File")]
            public AppenderFile File;
            [XmlElement("Pattern")]
            public AppenderPattern Pattern;
            [XmlElement("Color")]
            public List<AppenderColor> Colors;
        }

        public struct AppenderFile
        {
            [XmlAttribute]
            public string Path;
            [XmlAttribute]
            public string AppendTo;
        }

        public struct AppenderPattern
        {
            [XmlAttribute]
            public string Value;
        }

        public struct AppenderColor
        {
            [XmlAttribute]
            public string Level;
            [XmlAttribute]
            public string ForeColor;
            [XmlAttribute]
            public string BackColor;
        }

        public struct LogRoot
        {
            [XmlElement("MinLevel")]
            public RootMinLevel MinLevel;
            [XmlElement("AppenderRef")]
            public List<RootAppenderRef> RootAppenderRefs;
        }

        public struct RootMinLevel
        {
            [XmlAttribute]
            public string Value;
        }
        public struct RootAppenderRef
        {
            [XmlAttribute]
            public string Ref;
        }

        public void Load(string filePath)
        {
            XmlSerializer serializer = new XmlSerializer(GetType());
            LogConfig obj = Deserialize(serializer, filePath);
            if (null == obj || !CopyFrom(obj))
                throw new Exception("Failed to load data from XML: " + filePath);
        }

        public LogConfig Deserialize(XmlSerializer serializer, string filePath)
        {
            LogConfig obj = null;
            using (var stream = new FileStream(filePath, FileMode.Open, FileAccess.Read))
            {
                obj = (LogConfig)serializer.Deserialize(stream);
            }
            return obj;
        }

        public bool CopyFrom(LogConfig obj)
        {
            LogConfig source = obj as LogConfig;
            if (null == source) return false;
            if (null != source.Appenders)
                Appenders = source.Appenders;
            if (!source.Root.Equals(default(LogRoot)))
                Root = source.Root;
            return true;
        }
    }

}
